From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Novice-PG <Novice-PG@users.noreply.github.com>
Date: Sun, 04 Oct 2026 02:42:48 +0000
Subject: [PATCH] Add board KickPi K2B v2 DDR3 defconfig

KickPi K2B V2.2 is the DDR3 hardware revision. Reusing kickpi_k2b_defconfig
(LPDDR4 timings, DRAM_CLK 792) fails SPL DRAM training on this revision, so
the board hangs before the U-Boot prompt.

Add kickpi_k2b_v2_defconfig: SUNXI_DRAM_H616_DDR3_1333 at 648 MHz with H616
DDR3 timings validated on KickPi K2B V2.2 hardware (U-Boot prompt reached,
full boot). Board DTS stays shared with the original K2B. Derived from
Patrick Yavitz' "Add board KickPi K2B" board_kickpik2b/002 patch.

---
 configs/kickpi_k2b_v2_defconfig           |  35 +
 .../allwinner/sun50i-h618-kickpi-k2b.dts  | 327 +++++
 2 files changed, 362 insertions(+)
 create mode 100644 configs/kickpi_k2b_v2_defconfig
 create mode 100644 dts/upstream/src/arm64/allwinner/sun50i-h618-kickpi-k2b.dts

diff --git a/configs/kickpi_k2b_v2_defconfig b/configs/kickpi_k2b_v2_defconfig
new file mode 100644
index 00000000000..80ec70c3988
--- /dev/null
+++ b/configs/kickpi_k2b_v2_defconfig
@@ -0,0 +1,35 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="allwinner/sun50i-h618-kickpi-k2b"
+CONFIG_SPL=y
+CONFIG_DRAM_SUNXI_DX_ODT=0x03030303
+CONFIG_DRAM_SUNXI_DX_DRI=0x0e0e0e0e
+CONFIG_DRAM_SUNXI_CA_DRI=0x1f12
+CONFIG_DRAM_SUNXI_ODT_EN=0x0
+CONFIG_DRAM_SUNXI_TPR0=0xc0001002
+CONFIG_DRAM_SUNXI_TPR1=0x30fb
+CONFIG_DRAM_SUNXI_TPR2=0x0000
+CONFIG_DRAM_SUNXI_TPR6=0x33808080
+CONFIG_DRAM_SUNXI_TPR10=0x2f1107
+CONFIG_DRAM_SUNXI_TPR11=0xddddcccc
+CONFIG_DRAM_SUNXI_TPR12=0xeddc7665
+CONFIG_MACH_SUN50I_H616=y
+CONFIG_SUNXI_DRAM_H616_DDR3_1333=y
+CONFIG_DRAM_CLK=648
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
+CONFIG_R_I2C_ENABLE=y
+# CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+CONFIG_SPL_I2C=y
+CONFIG_CMD_KASLRSEED=y
+CONFIG_SPL_SYS_I2C_LEGACY=y
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_SYS_I2C_SLAVE=0x7f
+CONFIG_SYS_I2C_SPEED=400000
+CONFIG_SUPPORT_EMMC_BOOT=y
+CONFIG_SUN8I_EMAC=y
+CONFIG_AXP313_POWER=y
+CONFIG_AXP_DCDC3_VOLT=1500
+CONFIG_DM_RNG=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_MUSB_GADGET=y
diff --git a/dts/upstream/src/arm64/allwinner/sun50i-h618-kickpi-k2b.dts b/dts/upstream/src/arm64/allwinner/sun50i-h618-kickpi-k2b.dts
new file mode 100644
index 00000000000..e452a4a35fa
--- /dev/null
+++ b/dts/upstream/src/arm64/allwinner/sun50i-h618-kickpi-k2b.dts
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (C) 2025 Patrick Yavitz <pyavitz@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "sun50i-h616.dtsi"
+#include "sun50i-h616-cpu-opp.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/input/linux-event-codes.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "KickPi K2B";
+	compatible = "kickpi,k2b", "allwinner,sun50i-h618";
+
+	aliases {
+		ethernet0 = &emac0;
+		serial0 = &uart0;
+		serial5 = &uart5;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		poll-interval = <100>;
+
+		button-power {
+			label = "power";
+			linux,code = <KEY_POWER>;
+			gpios = <&pio 2 2 GPIO_ACTIVE_LOW>; /* PC2 */
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 8 16 GPIO_ACTIVE_HIGH>; /* PI16 */
+			linux,default-trigger = "heartbeat";
+		};
+	};
+
+	reg_usb_vbus: regulator-usb-vbus {
+		/* Separate discrete regulator for the USB ports */
+		compatible = "regulator-fixed";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-name = "usb-vbus";
+		vin-supply = <&reg_vcc5v>;
+	};
+
+	reg_vcc5v: regulator-vcc5v {
+		/* Board wide 5V supply directly from the USB-C socket */
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-name = "vcc-5v";
+	};
+
+	sdio_pwrseq: sdio-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		clocks = <&rtc CLK_OSC32K_FANOUT>;
+		clock-names = "ext_clock";
+		pinctrl-0 = <&x32clk_fanout_pin>;
+		pinctrl-names = "default";
+		post-power-on-delay-ms = <100>;
+		power-off-delay-us = <5000000>;
+		reset-gpios = <&pio 6 18 GPIO_ACTIVE_LOW>, /* PG18 */
+			      <&pio 6 19 GPIO_ACTIVE_LOW>; /* PG19 */
+	};
+};
+
+&cpu0 {
+	cpu-supply = <&reg_dcdc2>;
+};
+
+&emac0 {
+	status = "okay";
+	pinctrl-0 = <&ext_rgmii_pins>;
+	pinctrl-names = "default";
+	phy-mode = "rgmii";
+	phy-handle = <&ext_rgmii_phy>;
+	phy-supply = <&reg_dldo1>;
+	allwinner,tx-delay-ps = <700>;
+	allwinner,rx-delay-ps = <2000>;
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ehci3 {
+	status = "disabled";
+};
+
+&ir {
+	status = "okay";
+	pinctrl-0 = <&ir_rx_pin>;
+	pinctrl-names = "default";
+};
+
+&mdio0 {
+	ext_rgmii_phy: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0x0>;
+		reset-assert-us = <20000>;
+		reset-deassert-us = <100000>;
+		reset-gpios = <&pio 8 6 GPIO_ACTIVE_LOW>; /* PI6 */
+	};
+};
+
+/* SD card */
+&mmc0 {
+	status = "okay";
+	bus-width = <4>;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_HIGH>;	/* PF6 */
+	cd-inverted;
+	disable-wp;
+	max-frequency = <50000000>;
+	vmmc-supply = <&reg_dldo1>;
+};
+
+/* SDIO */
+&mmc1 {
+	status = "okay";
+	bus-width = <4>;
+	keep-power-in-suspend;
+	max-frequency = <100000000>;
+	mmc-pwrseq = <&sdio_pwrseq>;
+	no-mmc;
+	no-sd;
+	non-removable;
+	vmmc-supply = <&reg_dldo1>;
+	vqmmc-supply = <&reg_aldo1>;
+
+	sdio: wifi@1 {
+		reg = <1>;
+	};
+};
+
+/* eMMC */
+&mmc2 {
+	status = "okay";
+	bus-width = <8>;
+	cap-mmc-hw-reset;
+	mmc-hs200-1_8v;
+	non-removable;
+	vmmc-supply = <&reg_dldo1>;
+	vqmmc-supply = <&reg_aldo1>;
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
+&ohci3 {
+	status = "disabled";
+};
+
+&pio {
+	vcc-pc-supply = <&reg_dldo1>;
+	vcc-pf-supply = <&reg_dldo1>;
+	vcc-pg-supply = <&reg_aldo1>;
+	vcc-ph-supply = <&reg_dldo1>;
+	vcc-pi-supply = <&reg_dldo1>;
+	vcc-pl-supply = <&reg_aldo1>;
+
+	/* PI16 required for sys_led */
+	ext_rgmii_pins: rgmii-pins {
+		pins = "PI0", "PI1", "PI2", "PI3", "PI4",
+		       "PI5", "PI7", "PI8", "PI9", "PI10",
+		       "PI11", "PI12", "PI13", "PI14", "PI15";
+		function = "emac0";
+		drive-strength = <40>;
+	};
+
+	/* Add I2C PH Pins */
+	i2c1_ph_pins: i2c1-ph-pins {
+		pins = "PH0", "PH1";
+		function = "i2c1";
+	};
+
+	i2c2_ph_pins: i2c2-ph-pins {
+		pins = "PH2", "PH3";
+		function = "i2c2";
+	};
+
+	i2c4_ph_pins: i2c4-ph-pins {
+		pins = "PH6", "PH7";
+		function = "i2c4";
+	};
+
+	/* PC0 required for spi0 */
+	mmc2_pins: mmc2-pins {
+		pins = "PC1", "PC5", "PC6",
+		       "PC8", "PC9", "PC10", "PC11",
+		       "PC13", "PC14", "PC15", "PC16";
+		function = "mmc2";
+		drive-strength = <30>;
+		bias-pull-up;
+	};
+
+	/* Add UART Pins */
+	uart5_ph_pins: uart5-ph-pins {
+		pins = "PH2", "PH3";
+		function = "uart5";
+	};
+};
+
+&r_i2c {
+	status = "okay";
+	axp313: pmic@36 {
+		compatible = "x-powers,axp313a";
+		reg = <0x36>;
+		#interrupt-cells = <1>;
+		interrupt-controller;
+		interrupt-parent = <&pio>;
+
+		vin1-supply = <&reg_vcc5v>;
+		vin2-supply = <&reg_vcc5v>;
+		vin3-supply = <&reg_vcc5v>;
+
+		regulators {
+			reg_aldo1: aldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <1800000>;
+				regulator-step-delay-us = <25>;
+				regulator-final-delay-us = <50>;
+				regulator-name = "vcc-1v8-pll";
+			};
+
+			reg_dldo1: dldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-step-delay-us = <25>;
+				regulator-final-delay-us = <50>;
+				regulator-name = "vcc-3v3-io";
+			};
+
+			reg_dcdc1: dcdc1 {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <990000>;
+				regulator-step-delay-us = <25>;
+				regulator-final-delay-us = <50>;
+				regulator-name = "vdd-gpu-sys";
+			};
+
+			reg_dcdc2: dcdc2 {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <1100000>;
+				regulator-step-delay-us = <25>;
+				regulator-final-delay-us = <50>;
+				regulator-ramp-delay = <200>; /* FIXME */
+				regulator-name = "vdd-cpu";
+			};
+
+			reg_dcdc3: dcdc3 {
+				regulator-always-on;
+				regulator-min-microvolt = <1500000>;
+				regulator-max-microvolt = <1500000>;
+				regulator-step-delay-us = <25>;
+				regulator-final-delay-us = <50>;
+				regulator-name = "vdd-dram";
+			};
+		};
+	};
+};
+
+&uart0 {
+	status = "okay";
+	pinctrl-0 = <&uart0_ph_pins>;
+	pinctrl-names = "default";
+};
+
+/* Bluetooth */
+&uart1 {
+	status = "okay";
+	pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
+	pinctrl-names = "default";
+	uart-has-rtscts;
+};
+
+&uart5 {
+	status = "disabled";
+	pinctrl-0 = <&uart5_ph_pins>;
+	pinctrl-names = "default";
+};
+
+&usbotg {
+	status = "okay";
+	dr_mode = "peripheral";
+};
+
+&usbphy {
+	status = "okay";
+	usb1_vbus-supply = <&reg_usb_vbus>;
+};
-- 
2.43.0

